### PR TITLE
[ILM] Add unfollow action

### PR DESCRIPTION
**Pull request by martijnvg**
This change adds the unfollow action for CCR follower indices.

This is needed for the shrink action in case an index is a follower index.
This will give the follower index the opportunity to fully catch up with
the leader index, pause index following and unfollow the leader index.
After this the shrink action can safely perform the ilm shrink.

The unfollow action needs to be added to the hot phase and acts as
barrier for going to the next phase (warm or delete phases), so that
follower indices are being unfollowed properly before indices are expected
to go in read-only mode. This allows the force merge action to execute
its steps safely.

The unfollow action has three steps:
* `wait-for-indexing-complete` step: waits for the index in question
  to get the `index.lifecycle.indexing_complete` setting be set to `true`
* `wait-for-follow-shard-tasks` step: waits for all the shard follow tasks
  for the index being handled to report that the leader shard global checkpoint
  is equal to the follower shard global checkpoint.
* `unfollow-index` step: actually performs the unfollow. This consists
  out of multiple operations being executed on the index being handled:
  pause index following, close index, unfollow and open index. (a follower
  index can only be unfollowed when it is closed, because the underlying
  engine is changed)

In the case of the last two steps, if the index in being handled is
a regular index then the steps acts as a no-op.

Relates to #34648

### CODE REVIEW

File: x-pack/plugin/ilm/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/indexlifecycle/CCRIndexLifecycleIT.java
Comment (gwbrown) [inline x-pack/plugin/ilm/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/indexlifecycle/CCRIndexLifecycleIT.java]: This is a great first test, I think at a minimum we should have another one with an auto-follow pattern and have the index actually roll over and verify that everything goes as expected on the follower - we should be able to test that scenario as long as we don't shrink/delete the leader.

**Review by AthenaEryma — COMMENTED** [review]
Made some comments - some are questions, some are just to help track what still needs to be done with the understanding that this is still a WIP.

@martijnvg Don't worry about this until you get back - I'll work on this a bit like we discussed. I don't think this needs major changes, just a few tweaks/another integration test/etc.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/UnfollowFollowIndexStep.java
Comment (gwbrown) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/UnfollowFollowIndexStep.java]: One other thing: I'm not sure whether these are worth breaking out into their own AsyncActionSteps. That seems more in line with the intent of the model and make it clearer which part had a problem, but it might be worth it if it makes the interval we have to close the index for shorter. I'll investigate this.

File: x-pack/plugin/ilm/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/indexlifecycle/CCRIndexLifecycleIT.java
Comment (gwbrown) [inline x-pack/plugin/ilm/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/indexlifecycle/CCRIndexLifecycleIT.java]: I don't know what's up here, @martijnvg can we take a look at this when you're back?

Comment (martijnvg) [reply x-pack/plugin/ilm/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/indexlifecycle/CCRIndexLifecycleIT.java]: I pushed to following fix: 394944510484d31521f51333d3f31f5dc3fde650

The ccr read poll interval needed to be decreased.

The shard follow tasks continuesly poll changes from leader index,
in case no new writes arrive in the leader shards, then the leader
shards wait up to the defined read_poll_timeout before returning an
empty result. The result always includes meta information about whether
the index settings need to be synced.

Because in this test no data was indexed it took a full minute for ccr
to realize that the follower index settings needed to be updated and
caused the test to fail due to an assert busy statement to time out.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/WaitForFollowShardTasksStep.java
Comment (martijnvg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/WaitForFollowShardTasksStep.java]: If index following is paused then the follow stats api returns 0 stats entries.

> f it isn't, then shouldn't we put some other information in the Info to indicate what we're waiting for?

Yes, that makes sense.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/WaitForIndexingComplete.java
Comment (martijnvg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/WaitForIndexingComplete.java]: > Would you be okay with dropping the index settings from this Info and just keeping the message?

Yes, that makes sense. I was also thinking this before I went on vacation. I think I just forgot to change it to just keeping track of whether `index.xpack.ccr.following_index` setting is present.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/indexlifecycle/TimeseriesLifecycleTypeTests.java
Comment (martijnvg) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/indexlifecycle/TimeseriesLifecycleTypeTests.java]: I've adjusted the that test: 0ef70023b60e1e6986fdf4db0331f1df4f0c3682

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/UnfollowFollowIndexStep.java
Comment (martijnvg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/UnfollowFollowIndexStep.java]: I see, it is good to know at what part of the unfollow index step the policy is (pause index following, close index, unfollow or open index). Is there way to do that without adding a step for each action?

File: x-pack/plugin/ilm/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/indexlifecycle/CCRIndexLifecycleIT.java
Comment (gwbrown) [reply x-pack/plugin/ilm/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/indexlifecycle/CCRIndexLifecycleIT.java]: Thanks! Figured it was something like that, glad it was a simple fix.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/WaitForIndexingCompleteStep.java
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/WaitForIndexingCompleteStep.java]: For our other `ClusterStateWaitStep`s we are lenient with regard to the index no longer existing (being deleted in the state prior to this being run), can we make this one lenient as well?

For example:

```suggestion
        IndexMetaData followerIndex = clusterState.metaData().index(index);
        if (followerIndex == null) {
            // Index must have been since deleted, ignore it
            logger.debug("[{}] lifecycle action for index [{}] executed but index no longer exists", getKey().getAction(), index.getName());
            return new Result(false, null);
        }
```

**Review by dakrone — CHANGES_REQUESTED** [review]
I left some comments, thanks for doing this Martijn!

**Comment by martijnvg** [comment]
@dakrone @gwbrown I've split all operations but unfollow from `UnfollowFollowIndexStep` into their own index step implementation.

**Comment by dakrone** [comment]
Thanks @martijnvg, I notice this is marked as "WIP" is there more work that needs to go into this before another review?

Also, what do we think about having ILM wait for the index to be at least yellow before proceeding after opening the index in the unfollow? It might be nice to wait after that so the primaries are at least allocated before proceeding to any further steps in the ILM policy.

**Comment by gwbrown** [comment]
I would support waiting for the index to be at least yellow before continuing.

From my perspective, before final review, this needs:
1) a final decision on how exactly this action gets into a policy (is it implicit or explicit? If it's explicit this needs client support), which we should be addressing tomorrow, and 
2) docs, which we should also be ready to write after discussion tomorrow.

**Comment by martijnvg** [comment]
run the gradle build tests 1

**Comment by martijnvg** [comment]
I removed the WIP label as now all the required parts for the unfollow action are in this PR.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/WaitForYellowStep.java
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/WaitForYellowStep.java]: I think instead of making this an AsyncWaitStep that invokes a new ClusterHealthRequest, we should instead make it a `ClusterStateWaitStep` that checks it with:

```java
clusterState.routingTable().index(index).allPrimaryShardsActive()
```

I think this would be better so it doesn't invoke a new request, and should be simpler because there won't be any timeouts either

**Review by dakrone — CHANGES_REQUESTED** [review]
I left a few more comments, only one that requires a large change, but I think it will clean up the `WaitForYellowStep`

**Comment by martijnvg** [comment]
@dakrone Thanks for reviewing. I've updated the PR.

**Review by dakrone — APPROVED** [review]
This LGTM once CI is happy, thanks @martijnvg and @gwbrown

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/indexlifecycle/TimeseriesLifecycleTypeTests.java
Comment (gwbrown) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/indexlifecycle/TimeseriesLifecycleTypeTests.java]: I did not update this test when I added Unfollow to the Warm and Cold phases - I'll do that later today.

**Comment by gwbrown** [comment]
I spent a bit of time trying to track down the test failures, I saw a couple variations of failure but the one that reliably reproduces is that the test times out waiting for the `IndexRoutingTable` to become non-null, but I wasn't able to figure out why it never gets there.  I may be able to spend some more time on this later tonight, but wanted to write this up in case I can't.

**Comment by martijnvg** [comment]
The reason the qa test failed, was because of a silly bug: `OpenFollowerIndexStep` was extending from `AbstractUnfollowIndexStep` instead of just extending `AsyncActionStep`. Because of this, the open index api call was never made (the unfollow api had removed the custom ccr metadata).

**Comment by gwbrown** [comment]
As requested by @bmcconaghy, we're going to add the status of the `indexing_complete` setting to the StepInfo while waiting for it to be set as a boolean field, so that the UI can display it more easily.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/WaitForIndexingCompleteStep.java
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/WaitForIndexingCompleteStep.java]: I don't think we need to pass this through, it will always be false

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/WaitForIndexingCompleteStep.java
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/WaitForIndexingCompleteStep.java]: We could probably just hardcode `false` here? The class could be called `IndexingNotCompleteInfo` then too, since it's never used other than when the indexing is not complete

Comment (gwbrown) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/WaitForIndexingCompleteStep.java]: Fair, I went back and forth a little on whether to hardcode false or pass in the actual value. I'll switch it.

**Comment by martijnvg** [comment]
@elasticmachine run gradle build tests 2

**Comment by gwbrown** [comment]
@elasticmachine run gradle build tests 2

**Comment by martijnvg** [comment]
@elasticmachine run gradle build tests 1

**Comment by martijnvg** [comment]
@elasticmachine run gradle build tests 2

**Comment by gwbrown** [comment]
@elasticmachine run gradle build tests 2 please

**Pull request merged**
